### PR TITLE
crosscluster: rework how the change in plan is measured

### LINKED ISSUE #127022: ldr: job replanning doesn't replan on recovered nodes
In LDR resilience testing we found that when killing a node, the LDR job will not get replanned correctly to include the failed node (once it has recovered).  More details can be found [here](https://cockroachlabs.slack.com/archives/C077F53B4KY/p1720621620620169).

image

Jira issue: CRDB-40241

### PR DESCRIPTION
**Pull request by navsetlur**
The prior implementation of MeasurePlanChange was inaccurate at calculating the change between the current plan and a proposed new plan. This led to the scenario where a node being added to a cluster wouldn't be considered for any crosscluster replication. This change reworks the calculation to only count each unique node once and comparing the new plan count to the old one. If this value is seen to be higher than the threshold cluster setting, trigger an ingestion retry and subsequent replan.

Proof from roachprod:
image

Fixes: #127022
Release note: none

### CODE REVIEW

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 1]
```diff
@@ -509,6 +511,113 @@ func TestPreviouslyInterestingTables(t *testing.T) {
...
+		return plan
+	}
+	for _, tc := range []struct {
+		name   string
+		before sql.PhysicalPlan
+		after  sql.PhysicalPlan
+		frac   float64
+	}{
+		{
+			name:   "same node",
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:548]: what coverage is this adding? this looks essentially identical to TestMeasurePlanChange in stream_ingestion_dist_test.go

Instead, i'd add a regression test case to the original test.

**Comment by msbutler** [comment]
Also, mind ammending to the PR message to describe what the "proof from roachprod" is showing?

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go (hunk 1 above)
Comment (navsetlur) [reply pkg/ccl/crosscluster/logical/logical_replication_job_test.go]: I removed getNodes() from the test file so the test uses the real getNodes instead, which should cover any failures

Comment (msbutler) [reply pkg/ccl/crosscluster/logical/logical_replication_job_test.go]: aha, LDR and PCR have different getNodes implementations. This looks good to merge once you add a regression test case.

File: pkg/ccl/crosscluster/physical/stream_ingestion_dist_test.go [hunk 2]
```diff
@@ -54,25 +54,6 @@ func TestMeasurePlanChange(t *testing.T) {
 		return plan
 	}
 
-	getNodes := func(plan *sql.PhysicalPlan) (src, dst map[string]struct{}, nodeCount int) {
```
Comment (msbutler) [inline pkg/ccl/crosscluster/physical/stream_ingestion_dist_test.go:57]: in a follow up, non-backportable commit, we should move this test to the shared cross cluster package and test both the LDR and PCR getNodes implementations.

Comment (navsetlur) [reply pkg/ccl/crosscluster/physical/stream_ingestion_dist_test.go:57]: will do

File: pkg/ccl/crosscluster/physical/stream_ingestion_dist_test.go [hunk 3]
```diff
@@ -145,6 +126,12 @@ func TestMeasurePlanChange(t *testing.T) {
 			after:  makePlan(makeProc(1, []int{2}), makeProc(2, []int{1})),
 			frac:   0,
 		},
+		{
+			name:   "lots of processors",
```
Comment (msbutler) [inline pkg/ccl/crosscluster/physical/stream_ingestion_dist_test.go:130]: does this fail when your patch is commented out?

**Review by msbutler — APPROVED** [review]

File: pkg/ccl/crosscluster/physical/stream_ingestion_dist_test.go (hunk 3 above)
Comment (navsetlur) [reply pkg/ccl/crosscluster/physical/stream_ingestion_dist_test.go:130]: Yup

**Comment by navsetlur** [comment]
bors r=msbutler

**Comment by navsetlur** [comment]
blathers backport 24.2

**Comment by navsetlur** [comment]
blathers backport 24.1

**Pull request merged**